### ISSUE TITLE
Cannot upload images

### ISSUE THREAD

**Issue by paturdc** [inventree/InvenTree#1643]
Hi

I've managed to install and run the Inventree server software, and I'm delighted with the way it's working. Unfortunately, when I switched from Debug = True to Debug = False, I lost the ability to upload images. I'm using the default path '/home/inventree/data/media' in the config.yaml file. On the webserver interface the image seems to be located in http://*servername*/media/part_images/MF_RES.jpg but it doesn't show up. 

![image](https://user-images.githubusercontent.com/85507783/121050611-9ef59c00-c7b0-11eb-9d37-fe14708816aa.png)

It worked when running in Debug mode, but I've been unable to locate the files on the server.

**Comment by matmair** [comment]
@paturdc can you supply us with the version information (found in the about modal in the upper right corner of the page). Also do you use docker or setup manually?

**Comment by paturdc** [comment]
# Version Information:
InvenTree-Version: 0.2.3 pre
Django Version: 3.2.1
Commit Hash: 534fd49
Commit Date: 2021-06-07
Database: postgresql
Debug-Mode: False
Deployed using Docker: True

**Comment by matmair** [comment]
Docker is not my strong suit maybe @SchrodingersGat or @42CrMo4 has an idea.

**Comment by eeintech** [comment]
Looks like the setup for static/media files is not quite correct.

@paturdc You'll need to set your static and media folders, take a look in the doc: https://inventree.readthedocs.io/en/latest/start/config/#file-storage-locations

Then rerun `inv update` as it will copy the static files to the new folder.

Let us know if you have any more problem with that.

**Comment by paturdc** [comment]
No, that didn't work. The images work fine when setting Debug = True, and it seems it is only _loading_ that doesn't work when I set Debug = False. I tried uploading a new image while in Debug = False, and it did not show up. However, once I went back to debug mode, the new image was there. Like this:

Edit config.yaml -> Debug = False
Restart
Upload image -> broken link
Edit config.yaml -> Debug = True
Restart
Image visible

I don't think there is any problem with the static files. Also my config.yaml is located in this folder:

/var/lib/docker/volumes/docker_data/_data/config.yaml

which doesn't match the descriptions in the installation instructions. Inventree does read my changes to this file though.

**Comment by SchrodingersGat** [comment]
Hi @paturdc this is very strange.

Did you run through the recommended docker setup?

I'm going to try and replicate this behaviour. Can you share the docker-compose file you are using?

**Comment by SchrodingersGat** [comment]
@paturdc can you try doing a hard-refresh (Ctrl + Shift + R) and wee what happens?

I would like to know if you see any issues loading the javascript files

**Comment by matmair** [comment]
I have severall instances that run in production mode, either in 0.2.1 or 0.2.2 - they seems to be fine. But I have rather complicated hosting setups with authentication, proxies etc. and not docker.

**Comment by SchrodingersGat** [comment]
Upon further reading I've stumbled across the answer - in production mode, the following line **does not work**:

https://github.com/inventree/InvenTree/blob/fd63e49e60710a909c64aed1a770e12248c9863b/InvenTree/InvenTree/urls.py#L168

Specifically, the `static(...)` call does not work in when `debug=False`

![image](https://user-images.githubusercontent.com/10080325/121181232-202e5b00-c8a5-11eb-8005-a2db9e9c9623.png)

In the docker configuration, the static files are served by the nginx container, and so the InvenTree wsgi application never sees the requests to `/static/`

However, requests to `/media/` are still getting through.

@matmair what would you suggest here? Currently the docker build relies entirely on `debug=True` being set in the config file - clearly not best practice!

Simple option might be to allow the InvenTree wsgi server (via gunicorn) to handle media file serving, which is not ideal but probably easier than configuring a separate service to host the media files.

We *could* configure the nginx container to serve media files, but the big unknown for me there is managing permissions. If the nginx server is simply "hosting" the media files, couldn't anyone upload / delete / tamper with media files without being authenticated? The InvenTree server already handles that..

**Comment by paturdc** [comment]
So the strange path behaviour came about because I don't know what Docker is, and I think I made a 'mixed' installation running some elements of the manual installation first, before docker-compose.

Anyway, the webserver is running now, although the image issue persists. Images are uploaded in both debug and production mode, placed correctly in the `data/media/part_images directory`, but cannot be viewed on the webpage when running in production mode.

Are there reasons not to run in debug mode in general?

**Comment by matmair** [comment]
@eeintech  that is simpel with apache, i am not sure aboug nginx. We already have nginx running so we could mount static and media from there and pass permissions through to InvenTree.
[Apache](https://docs.djangoproject.com/en/3.2/howto/deployment/wsgi/apache-auth/)
[Nginx](https://stackoverflow.com/questions/46421589/nginx-location-and-django-auth) [Official doc](https://docs.nginx.com/nginx/admin-guide/security-controls/configuring-subrequest-authentication/)  -> [Blog that shows example usage](https://pawamoy.github.io/posts/django-auth-server-for-shiny/)
Nginx generally does not allow manipulation of the files served, so just hosting so that would be fine.

**Comment by matmair** [comment]
@paturdc the doc is howsted [here](https://github.com/inventree/inventree-docs) feel free to submit a PR

**Comment by eeintech** [comment]
For my company I'm running the latest from master in production mode without any issue, I even uploaded a picture yesterday, works completely fine. Both static and media files are serve correctly.
I have a rather different setup, using Phusion Passenger (standalone) as our web server, works great with Django.

@paturdc Maybe run through the deployment checklist to see what you come up with?
https://docs.djangoproject.com/en/3.2/howto/deployment/checklist/

**Comment by SchrodingersGat** [comment]
@paturdc for now I'd suggest setting `debug=True` in your configuration file, until we can work out how to fix the docker issues when running in production mode.

> I have some suggestions how to make the Docker Installation instructions a little easier for people like me to follow, if you wish.

I'd love to hear them! You can submit a documentation PR and we can discuss there.

**Comment by SchrodingersGat** [comment]
@matmair sounds like there is definitely support for nginx to handle media requests, then.

I'll have to have a read through that documentation link you shared, and see how we can add a /media section to the nginx.conf file

**Comment by matmair** [comment]
@SchrodingersGat I am a Apache2 guy (except for a simple proxy) but I have legacy nginx configs that do something similar.

**Comment by paturdc** [comment]
@SchrodingersGat I only have the most superficial understaning of what is happening behind the scenes with Inventree. As I understand it, the reason some people see their media is because they are using the manual install, and with Docker there is a need to load data between containers and outside folders? And the code for doing so relies on `Debug = true ` being set in Django?

Anyway, the fact that I can access the media directory in Debug mode and not in production, seems to indicate that it is not a problem with my path setup. The line `path/to/data` at the end of docker-compose.yml I took to refer to any path of my own choosing outside the container. And the line `/home/inventree/media` in config.yaml I've understood to be linked to the chosen `path/to/data` by some programming magic.

I'm running this on a webserver (it's a learning exercise), so it might not be a good idea to set debug = true. I can live without images for a while, if I'm confident my installation if otherwise correct.

**Comment by nwns** [comment]
When Django is run in Debug mode, the server serves media and static files itself to make development easier, but when in production mode it does not.

Your paths are probably fine.

I have not been able to test it, but adding the following to the nginx.conf should enable it to serve the media files.

```
    
    # Redirect any requests for media files
    location /media/ {
        alias /home/inventree/data/media/;

        # Caching settings
        expires 1d;
        add_header Pragma private;
        add_header Cache-Control "private";
    }
```

**Comment by paturdc** [comment]
@nwns That makes a lot of sense.

Before this change, clicking on the image path on the website would just lead me back to index. After the change, the webserver is directed to `http://ip-address/media/part_images/MF_RES.jpg` although rather than an image I get the Nginx 404 page. Gotta look at it again after the weekend. Thanks everybody for all the input so far.

**Comment by paturdc** [comment]
So I managed to make this work, by changing the line in @nwns suggestion to `alias /var/www/data/media/;`

Then in the docker-compose file I added the following lines under the inventree-proxy:

`# Media data volume is mounted to /var/www/data
- data:/var/www/data:rw`

then restarting. I think that's everything. Someone with more experience could tell me if some part of this is bad practice or not secure.

**Comment by SchrodingersGat** [comment]
@paturdc thanks for the feedback, I'll try and use this information to fix the issue :)

**Comment by SchrodingersGat** [comment]
> 
> 
> So I managed to make this work, by changing the line in @nwns suggestion to `alias /var/www/data/media/;`
> 
> Then in the docker-compose file I added the following lines under the inventree-proxy:
> 
> `# Media data volume is mounted to /var/www/data`
> `- data:/var/www/data:ro`
> 
> then restarting. I think that's everything. Someone with more experience could tell me if some part of this is bad practice or not secure.

Have you tested file uploads (e.g. setting a part image) with this configuration?

**Comment by paturdc** [comment]
@SchrodingersGat yes, I tried it just now. And in any case uploading seemed to not be a problem as I could always locate the files locally on the server. They were just never shown on the website.

**Comment by matmair** [comment]
@paturdc I could see that one can access the media without beeing authenticated with this config. You might want to check if that is the case. Could be quite bad depending on the kind of your install.

**Comment by paturdc** [comment]
@matmair that is indeed the case. For what it is worth, this is also true of static files:

![image](https://user-images.githubusercontent.com/85507783/122181052-e7801a00-ce80-11eb-8cdd-92c3a648c3c9.png)

and I haven't changed any of that configuration, to my knowledge.

What is the simple solution, authentication in nginx perhaps?

**Comment by matmair** [comment]
> 
> 
> @matmair that is indeed the case. For what it is worth, this is also true of static files:
> 
> and I haven't changed any of that configuration, to my knowledge.

static is no problem as there is no user-generated data in there. Those are same for all deployments and do not reveal anything.
> 
> What is the simple solution, authentication in nginx perhaps?

Authentication would be the simplest. I personally have everything forwarded through a proxy that checks against SAML / Kerberos that the user is in the right group and on the right endpoint.

**Comment by SchrodingersGat** [comment]
I have a working example here - https://github.com/inventree/InvenTree/pull/1664

Basically this serves /static files without auth, but requires auth for /media files.

Currently tweaking the dockerfile (etc) so this works "out of the box"

**Comment by matmair** [comment]
@SchrodingersGat nice clean solution!

**Comment by SchrodingersGat** [comment]
@paturdc once the docker image for 0.2.3 has been built and uploaded, please try again and let me know if it works for you now :)
